OpenTrons/opentrons-api
# feat(protocol-designer): deck setup tools

closes AUTh-673

# Overview

Create the deck setup tools for selecting labware and modules and fixtures to go on a particular slot. Note that deck map zoom is not wired up, this is strictly for selecting items from the toolbox.


https://github.com/user-attachments/assets/1f1c842d-0590-4d80-b4d7-3a10d4453418


## Test Plan and Hands on Testing

There are a bunch of things we should test:

1. turn on the redesign ff and create a flex protocol. Navigate to the deck map view in the redesign
- click on slot B1, see that all modules are available and Trash bin is available but greyed out, all labware for going on a slot should be available
- click on slot D3, see that all modules except TC are available and all fixtures should be available. All labware for going on a slot should be available
- click on a middle slot, see that only the magnetic block is available. Add the magnetic block and add a labware to it and see that only the magnetic block supported labware are available
- now click on a slot where a heater-shaker can go into, see that only adapters are available and selecting on one should open the nested accordion to add another labware on the adapter. Hit Done and it should generate all entities.
- now click on a staging area slot and add a staging area. go to the labware and see all labware for going on a slot is avaiable
- finally, click clear, see that everything selected is reset and slot should be empty too

2. now test out the 96-channel specific stuff. in order to do this, turn on the ff and add a 96-channel the current way. then turn the ff back on and go to the deck setup, should see a labware on top an adapter for the tiprack/adapter combo. Click to add a tiprack and see that the 96-channel tiprack(s) you added are available

3. test the ot-2. See that the TCs are only available for slot 8 but all modules are available for all outer slots and no modules are available for the inner slots

4. test the custom labware functionality. test uploading a custom labware and see that it shows correctly and is selectable. (I've attached a custom labware)

[shlok_64_tiprack_20ul.json](https://github.com/user-attachments/files/16659396/shlok_64_tiprack_20ul.json) 

## Changelog

- create `DeckSetupTools` and `LabwareTools` - basically the redesign version of `LabwareSelectionModal`
- add tests
- organize the utils, constants, and types
- create new thunk for generating labware on top of a module (thunk is to grab the state to get the `moduleId` to generate the labware's slot)
- add some test labware fixtures for testing an adapter and labware on an adapter

## Review requests

see test plan

## Risk assessment

low
